### DISCUSSION TITLE
MerkleAirdrop test failing.

### DISCUSSION THREAD

**Discussion by ArlisVilipold** [Cyfrin/foundry-full-course-cu#2402] [Q&A]
`Ran 1 test for test/MerkleAirdrop.t.sol:MerkleAirdropTest
[FAIL. Reason: MerkleAirdrop__InvalidProof()] testUsersCanClaim() (gas: 30546)
Traces:
  [30546] MerkleAirdropTest::testUsersCanClaim()
    ├─ [2563] BagelToken::balanceOf(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D]) [staticcall]
    │   └─ ← [Return] 0
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [3801] MerkleAirdrop::claim(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 25000000000000000000 [2.5e19], [0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a, 0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576])
    │   └─ ← [Revert] MerkleAirdrop__InvalidProof()
    └─ ← [Revert] MerkleAirdrop__InvalidProof()

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 553.34µs (79.91µs CPU time)

Ran 1 test suite in 3.22ms (553.34µs CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/MerkleAirdrop.t.sol:MerkleAirdropTest
[FAIL. Reason: MerkleAirdrop__InvalidProof()] testUsersCanClaim() (gas: 30546)

Encountered a total of 1 failing tests, 0 tests succeeded`

`// SPDX-License-Identifier: MIT

pragma solidity ^0.8.24;

import {Test, console} from "forge-std/Test.sol";
import {MerkleAirdrop} from "../src/MerkleAirdrop.sol";
import {BagelToken} from "../src/BagelToken.sol";
import {ZkSyncChainChecker} from "lib/foundry-devops/src/ZkSyncChainChecker.sol";
import {DeployMerkleAirdrop} from "script/DeployMerkleAirdrop.s.sol";

contract MerkleAirdropTest is ZkSyncChainChecker, Test {

    MerkleAirdrop public airdrop;
    BagelToken public token;

    bytes32 public ROOT = 0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4;
    address user;
    uint256 userPrivKey;
    uint256 public AMOUNT_TO_CLAIM = 25 * 1e18;
    uint256 public AMOUNT_TO_SEND = AMOUNT_TO_CLAIM * 4;
    bytes32 proofOne = 0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a;
    bytes32 proofTwo = 0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576;
    bytes32[] public PROOF = [proofOne, proofTwo];

  
    bytes32[] proof = [proofOne, proofTwo];

    function setUp() public {
        if (!isZkSyncChain()){
            DeployMerkleAirdrop deploy = new DeployMerkleAirdrop();
           (airdrop, token) = deploy.deployMerkleAirdrop();  
        } else {
            token = new BagelToken();
        airdrop = new MerkleAirdrop(ROOT, token);
        token.mint(token.owner(), AMOUNT_TO_SEND);
        token.transfer(address(airdrop), AMOUNT_TO_SEND);
        }
        (user, userPrivKey) = makeAddrAndKey("user");
    }

    function testUsersCanClaim() public {
        uint256 startingBalance = token.balanceOf(user);

        vm.prank(user);
        airdrop.claim(user, AMOUNT_TO_CLAIM, PROOF);

        uint256 endingBalance = token.balanceOf(user);
        console.log("Ending baalance:", endingBalance);
        assertEq(endingBalance - startingBalance, AMOUNT_TO_CLAIM);

    }
}`

`[
    {
        "inputs": [
            "0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D",
            "25000000000000000000"
        ],
        "proof": [
            "0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a",
            "0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576"
        ],
        "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4",
        "leaf": "0xd1445c931158119b00449ffcac3c947d028c0c359c34a6646d95962b3b55c6ad"
    },
    {
        "inputs": [
            "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
            "25000000000000000000"
        ],
        "proof": [
            "0xd1445c931158119b00449ffcac3c947d028c0c359c34a6646d95962b3b55c6ad",
            "0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576"
        ],
        "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4",
        "leaf": "0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a"
    },
    {
        "inputs": [
            "0x2ea3970Ed82D5b30be821FAAD4a731D35964F7dd",
            "25000000000000000000"
        ],
        "proof": [
            "0x4fd31fee0e75780cd67704fbc43caee70fddcaa43631e2e1bc9fb233fada2394",
            "0x81f0e530b56872b6fc3e10f8873804230663f8407e21cef901b8aeb06a25e5e2"
        ],
        "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4",
        "leaf": "0x0c7ef881bb675a5858617babe0eb12b538067e289d35d5b044ee76b79d335191"
    },
    {
        "inputs": [
            "0xf6dBa02C01AF48Cf926579F77C9f874Ca640D91D",
            "25000000000000000000"
        ],
        "proof": [
            "0x0c7ef881bb675a5858617babe0eb12b538067e289d35d5b044ee76b79d335191",
            "0x81f0e530b56872b6fc3e10f8873804230663f8407e21cef901b8aeb06a25e5e2"
        ],
        "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4",
        "leaf": "0x4fd31fee0e75780cd67704fbc43caee70fddcaa43631e2e1bc9fb233fada2394"
    }
]`

the roots and proofs are correct imo.. but still is failing .

**Comment by EngrPips** [comment]
```js
[ { "inputs": [ "0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D", "25000000000000000000" ], "proof": [ "0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a", "0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576" ], "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4", "leaf": "0xd1445c931158119b00449ffcac3c947d028c0c359c34a6646d95962b3b55c6ad" }, { "inputs": [ "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266", "25000000000000000000" ], "proof": [ "0xd1445c931158119b00449ffcac3c947d028c0c359c34a6646d95962b3b55c6ad", "0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576" ], "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4", "leaf": "0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a" }, { "inputs": [ "0x2ea3970Ed82D5b30be821FAAD4a731D35964F7dd", "25000000000000000000" ], "proof": [ "0x4fd31fee0e75780cd67704fbc43caee70fddcaa43631e2e1bc9fb233fada2394", "0x81f0e530b56872b6fc3e10f8873804230663f8407e21cef901b8aeb06a25e5e2" ], "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4", "leaf": "0x0c7ef881bb675a5858617babe0eb12b538067e289d35d5b044ee76b79d335191" }, { "inputs": [ "0xf6dBa02C01AF48Cf926579F77C9f874Ca640D91D", "25000000000000000000" ], "proof": [ "0x0c7ef881bb675a5858617babe0eb12b538067e289d35d5b044ee76b79d335191", "0x81f0e530b56872b6fc3e10f8873804230663f8407e21cef901b8aeb06a25e5e2" ], "root": "0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4", "leaf": "0x4fd31fee0e75780cd67704fbc43caee70fddcaa43631e2e1bc9fb233fada2394" } ]
```

**Reply by ArlisVilipold** [reply]
still invalid proof

**Reply by EngrPips** [reply]
Please show your `DeployMerkleAirdrop`

**Reply by ArlisVilipold** [reply]
//SPDX-License-Identifier: MIT

pragma solidity ^0.8.24;

import {MerkleAirdrop} from "../src/MerkleAirdrop.sol";
import {BagelToken} from "../src/BagelToken.sol";
import {Script} from "forge-std/Script.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";

contract DeployMerkleAirdrop is Script {

    bytes32 private s_merkleRoot = 0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4;
    uint256 private s_amountToTransfer = 4 * 25 * 1e18;

    function deployMerkleAirdrop() public returns (MerkleAirdrop, BagelToken) {
        vm.startBroadcast();
        BagelToken token = new BagelToken();
        MerkleAirdrop airdrop = new MerkleAirdrop(s_merkleRoot, IERC20(address(token)));
        token.mint(token.owner(), s_amountToTransfer);
        token.transfer(address(airdrop), s_amountToTransfer);
        vm.stopBroadcast();
        return (airdrop, token);
    }
    function run() external returns (MerkleAirdrop, BagelToken) {
        return deployMerkleAirdrop();
    }

}

**Reply by EngrPips** [reply]
Everything seems fine, hmmm 🤔 Please run `forge clean` and then retry the test. Let's see if that helps.

**Reply by ciaranightingale** [reply]
hey @ArlisVilipold is this still an issue? If so, can I see your `MerkleAirdrop.sol` file please

**Reply by EngrPips** [reply]
> hey @ArlisVilipold is this still an issue? If so, can I see your `MerkleAirdrop.sol` file please

Hello @IshaanXCoder, Maybe provide your `MerkleAirdrop.sol` as @ciaranightingale mentioned.

**Comment by Lamma-maihadisi** [comment]
i have same issue, pls @EngrPips  is there any alternative solution ? thank you

**Reply by EngrPips** [reply]
Please show your test stack trace and format it nicely. Thanks

**Comment by IshaanXCoder** [comment]
heyy @Lamma-maihadisi @ArlisVilipold is this is solved? I'm facing the same issue, pls lemme know hot to fix this up

**Reply by AlesxanDer1102** [reply]
Check the function claim in `MerkleAirdrop.sol` I have the same issue and I change `bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encodePacked(account, amount))));` to `bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(account, amount))));` and It works good

**Reply by EngrPips** [reply]
Awesome! Thanks for the heads-up.
